lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3Context.java

```diff
@@ -22,15 +22,20 @@
 import software.amazon.awssdk.awscore.AwsRequestOverrideConfiguration;
 import software.amazon.awssdk.services.s3.model.RequestPayer;
 
+import java.util.Base64;
 import java.util.Optional;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.trino.filesystem.s3.S3FileSystemConfig.S3SseType.CUSTOMER;
 import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_ACCESS_KEY_PROPERTY;
 import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_SECRET_KEY_PROPERTY;
 import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_SESSION_TOKEN_PROPERTY;
+import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_SSEC_KEY;
 import static java.util.Objects.requireNonNull;
+import static software.amazon.awssdk.utils.Md5Utils.md5AsBase64;
 
-record S3Context(int partSize, boolean requesterPays, S3SseType sseType, String sseKmsKeyId, Optional<AwsCredentialsProvider> credentialsProviderOverride, ObjectCannedAcl cannedAcl, boolean exclusiveWriteSupported)
+record S3Context(int partSize, boolean requesterPays, S3SseType sseType, String sseKmsKeyId, Optional<AwsCredentialsProvider> credentialsProviderOverride, Optional<String> encryptionKey, ObjectCannedAcl cannedAcl, boolean exclusiveWriteSupported)
```

Comment:
The constructor needs to validate that `encryptionKey` is only set when `sseType` is `CUSTOMER`, and not set for other SSE types.

Reply:
Thinking about this more, since this SSE-C support is only for internal use, let's enforce that the encryption key is only set when `sseType` is `NONE`. This will prevent users from setting it to `CUSTOMER`, as the enum is used for the user configuration.